### PR TITLE
[PyTorch] Fix NaN expert weight gradients at num_groups == 1 with SReLU

### PR DESCRIPTION
**Pull request by GarlGuo**
# Description

At `num_groups == 1` the fused grouped MLP may quantize and scale-swizzle the token buffer as one dense tensor over `tensor.shape[0]` rows and read it back the same way. That is only coherent if the cuDNN kernels *also* write their intermediates densely over `tensor.shape[0]`, which is what `use_single_group_runtime_offsets` asks of them — they then derive M from the runtime tensor shapes rather than from the offsets.

TE selects the dense consumer without checking whether the producer will be dense. The two disagree for `ScaledSReLU` at any version, and for every activation on a cuDNN frontend older than 1.27.0, because `_cudnn_frontend_supports_single_group_runtime_offsets` returns False there and the kernels pack their output to `sum(split_sizes)` rows instead.

The columnwise swizzled MXFP8 scale layout is `[k/128][m/128][32][4][4]`, with the m-tile as an **inner** stride, so an m-extent mismatch misindexes every k-tile past the first: consumer k-tile `j` reads producer k-tile `j * T_c / T_p`. The head rows — the ones that *do* belong to the group — pick up the wrong scale factors, and high k-tiles read scale memory the producer never wrote. The data is read correctly; only the scales move, so the error is order 1 and often NaN.

Crucially this is **value-independent**: zeroing the padded tail does not help, because the defect is in the indexing rather than in the padding's contents. A caller cannot work around it.

Callers may legally pass a token buffer padded past `sum(split_sizes)`; an MoE expert-parallel fixed-capacity receive buffer always does, and `test_grouped_linear_cuda_graph_safe` documents the padded tail as "intentionally outside every group" and "uninitialized".

Fix to #3484 

### Minimal reproduction

Single GPU, no Megatron, no distributed. `NVTE_CUTEDSL_FUSED_GROUPED_MLP=1` must be set before the TE import.

```python
import os
os.environ["NVTE_CUTEDSL_FUSED_GROUPED_MLP"] = "1"  # must precede the TE import
import torch
import transformer_engine.pytorch as te
from transformer_engine.pytorch.quantization import MXFP8BlockScaling

HIDDEN, LIVE, PAD = 1024, 512, 256


def wgrad(total_rows, fill):
    torch.manual_seed(0)
    fc1 = te.ops.GroupedLinear(1, HIDDEN, HIDDEN, bias=False, device="cuda",
                               dtype=torch.bfloat16, accumulate_into_main_grad=True)
    fc2 = te.ops.GroupedLinear(1, HIDDEN, HIDDEN, bias=False, device="cuda",
                               dtype=torch.bfloat16, accumulate_into_main_grad=True)
    model = te.ops.Sequential(fc1, te.ops.ScaledSReLU(), fc2)
    for op in (fc1, fc2):
        w = op.get_parameter("weight0")
        w.main_grad = torch.zeros(w.shape, dtype=torch.float32, device="cuda")

    gen = torch.Generator(device="cuda").manual_seed(1)
    live_x = torch.randn(LIVE, HIDDEN, generator=gen, device="cuda", dtype=torch.bfloat16)
    live_dy = torch.randn(LIVE, HIDDEN, generator=gen, device="cuda", dtype=torch.bfloat16)

    x = torch.full((total_rows, HIDDEN), fill, device="cuda", dtype=torch.bfloat16)
    dy = torch.full((total_rows, HIDDEN), fill, device="cuda", dtype=torch.bfloat16)
    x[:LIVE], dy[:LIVE] = live_x, live_dy
    x = x.detach().requires_grad_(True)

    # split_sizes covers the LIVE rows only; the rest of the buffer is padding.
    splits = torch.tensor([LIVE], device="cuda", dtype=torch.int32)
    probs = torch.ones(total_rows, device="cuda", dtype=torch.float32)

    with te.autocast(enabled=True, recipe=MXFP8BlockScaling()):
        out = model(x, splits, probs, splits)
    out.backward(dy)
    return fc1.get_parameter("weight0").main_grad.double()


reference = wgrad(LIVE, 0.0)                      # no padding
for label, fill in (("padding zeroed  ", 0.0), ("padding garbage ", 3.0)):
    got = wgrad(LIVE + PAD, fill)
    rel = ((got - reference).norm() / reference.norm()).item()
    bad = int((~torch.isfinite(got)).sum())
    print(f"{label} rel_err={rel:.4e}  non_finite={bad}/{got.numel()}")
```

On `main`:

```
padding zeroed   rel_err=nan  non_finite=411648/1048576
padding garbage  rel_err=nan  non_finite=411648/1048576
```

With this PR:

```
padding zeroed   rel_err=0.0000e+00  non_finite=0/1048576
padding garbage  rel_err=0.0000e+00  non_finite=0/1048576
```

Note the two `main` rows are identical: the corruption does not depend on what the padding holds, so a caller cannot avoid it by zeroing.

### Why the fix is where it is

The two halves of the "treat this as one dense group" decision are currently gated differently. The request to the kernels is already guarded, because it can be refused:

```python
if supports_single_group_runtime_offsets:
    fc1_activation_kwargs["use_single_group_runtime_offsets"] = num_groups == 1
```

but its siblings — `use_single_group_weight_swizzle`, `use_single_discrete_weight`, `use_single_group_dense_fc2`, `use_single_group_dense_dgrad`, the dense quantize and the plain wgrad GEMM — are gated on `num_groups == 1` alone. So when the specialization is unavailable, TE still reads everything densely while the kernels pack.

This PR applies the guard TE already writes, at the sites that were missing it:

```python
use_dense_single_group = num_groups == 1 and (
    _cudnn_frontend_supports_single_group_runtime_offsets(type(activation_op))
)
```

That predicate is called directly rather than wrapped, matching the three other call sites in this file, and the flag is named for the decision it drives rather than for a property of the data. It deliberately does not claim the single group spans the buffer — `split_sizes` is device-resident, so that is unknowable here — only that the producer and the consumer will agree on the row extent.

They have to move together: quantization layout and kernel indexing must agree, so enabling a subset pairs a dense consumer with a packed producer and yields NaN rather than a slowdown. Hence one flag threaded explicitly through the grouped-quantize helpers and `_compute_grad_params`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Gate the `num_groups == 1` dense shortcut family on `_cudnn_frontend_supports_single_group_runtime_offsets(type(activation_op))` as well as on the group count, so TE only reads operands densely when the cuDNN kernels also write them densely.
- Thread that one decision explicitly through `_group_quantize_for_grouped_mlp`, `_group_quantize_with_amax_for_grouped_mlp` and `_compute_grad_params` rather than recomputing it per site, so all of the shortcuts move together.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

`tests/pytorch/test_grouped_mlp.py` is unmodified and at baseline: 6 failed, 34 passed, the 6 being pre-existing `nvfp4_rht` and cuBLAS-version failures also present on `main`. The fix is verified out-of-tree by the reproducer above; happy to add an in-tree regression test if you would like one in this PR.

### CODE REVIEW

File: transformer_engine/pytorch/ops/fused/grouped_mlp.py [hunk 1]
```diff
@@ -239,10 +195,15 @@ def _group_quantize_for_grouped_mlp(
...
+    use_dense_single_group: bool = False,
 ) -> GroupedTensor:
     """Quantize into grouped storage."""
 
-    if num_groups != 1 or not isinstance(quantizer, (MXFP8Quantizer, NVFP4Quantizer)):
+    if (
+        num_groups != 1
+        or not use_dense_single_group
+        or not isinstance(quantizer, (MXFP8Quantizer, NVFP4Quantizer))
+    ):
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/ops/fused/grouped_mlp.py]:
```suggestion
    if  not use_dense_single_group:
```

We only support mxfp8/nvfp4 quantizer in grouped_mlp. So checking quantizer seems not relavant to me. So we should remove it. cc: @sraman-rgb

File: transformer_engine/pytorch/ops/fused/grouped_mlp.py [hunk 2]
```diff
@@ -279,9 +241,10 @@ def _group_quantize_with_amax_for_grouped_mlp(
             num_groups,
             split_sizes,
             tensor_offsets=tensor_offsets,
+            use_dense_single_group=use_dense_single_group,
         )
 
-    if num_groups != 1:
+    if num_groups != 1 or not use_dense_single_group:
```
Comment (vthumbe1503) [inline transformer_engine/pytorch/ops/fused/grouped_mlp.py]:
```suggestion
    if  not use_dense_single_group:
```

num_groups condition is already encoded in use_dense_single_group

**Review by vthumbe1503 — COMMENTED** [review]
Fix LGTM. Have minor comments.

**Comment by GarlGuo** [comment]
@vthumbe1503 could you take another pass? thank you!

**Review by vthumbe1503 — APPROVED** [review]
LGTM. Pending CI
